src/math/interpolator.hh

```diff
@@ -0,0 +1,1649 @@
+ *
+ * rascal is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation, either version 3, or (at
+ * your option) any later version.
+ *
+ * rascal is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this software; see the file LICENSE. If not, write to the
+ * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
+ * Boston, MA 02111-1307, USA.
+ */
+
+#ifndef SRC_MATH_INTERPOLATOR_HH_
+#define SRC_MATH_INTERPOLATOR_HH_
+
+#include <functional>
+#include <forward_list>
+#include <iostream>
+#include <limits>
+#include <cassert>
+#include "math_utils.hh"
+
+namespace rascal {
+  namespace math {
+    /**
+     * To allow flexibility for the behaviour of the interpolator for
+     * experimenting while giving the possibility to optimize the interpolator
+     * for certain methods, the interpolator class itself is only a skeleton to
+     * execute these interpolator methods.
+     *
+     * There exist methods for the mathematical derived interpolation method
+     * contained in InterpolationMethod, for the creation of the grid and test
+     * grid contained in the GridRational, a search method to find the nearest
+     * grid point before a point for interpolation contained in SearchMethod and
+     * a method for computing an error term contained in ErrorMethod.
+     *
+     * The Interpolator class hierarchy is are split into InterpolatorBase ->
+     * Interpolator (abstract) -> YOUR_INTERPOLATOR_IMPLEMENTATION An
+     * interpolator implementation can be very general an allow multiple
+     * combination of methods to allow flexibility or very rigid to allow better
+     * optimization. The same goes for the interpolator methods which can be
+     * used interchangeable with other methods or depend that certain other
+     * methods are used.
+     *
+     * While testing different methods it has turned out that the interpolator
+     * is most optimal for uniform grids, therefore there exists only
+     * implementations using uniform grids and interpolation methds optimized
+     * for uniform non adaptive grids. However, the adaptive GridRational and
+     * SearchMethod for non uniform grids are functional for an appropriate
+     * interpolator implementation and are kept in case of need.
+     */
+
+    /**
+     * GridRational consist of an refinement part which determines the method of
+     * refining the grid. The GridType describes the type of starting grid, but
+     * the type does necessery have to hold after the refinement.
+     */
+    enum class GridType_t { Uniform };
```

Comment:
What if you want the option to add other grid types in the future?

Reply:
Then add a new enum in the future ?

Reply:
Having a single use enum adds to code clutter and make it less readable. At the same time, the whole Interpolator class assumes uniform grids anyway, so the amount of work needed to define a new enum is negligible compared to the actual implementation effort.

Also, YAGNI: keep it as simple as possible until more is needed.

Reply:
I understand the reasoning (and the YAGNI argument), but there's a difference between implementing a useless feature and implementing something which might have some legitimate scientific use, so of course we want to leave that option open (and not bother people with "oh, but you have to extend such-and-such a class with this enum, and then change this and that, and then do this to make it compile..." just to make a _scientific_ comparison between two legitimate alternatives).

For example: In the spherical expansion class, we have an enum to select between how the atom smearings are implemented (constant, radially-dependent, or species-dependent?). Currently the constant version is the only thing that's implemented, but you can't just YAGNI away the other two versions just because we haven't implemented them yet. They have a legitimate scientific purpose that we want to make available to non-developers in the future.

Granted, this seems like a slightly different case, if (as you say) most of the implementation is already specialized to the uniform-grid case. But let's be a little less aggressive in applying the YAGNI principle, because one of the main goals of this code is to give scientists as much flexibility in playing around with descriptor parameters as we can practically do.

Reply:
>  because one of the main goals of this code is to give scientists as much flexibility in playing around with descriptor parameters as we can practically do.

I think we agree on the goals, but not on how to achieve them. The code already feels very complex, and making it simpler would enable more users to play with it, without them needing to be C++ experts.

> For example: In the spherical expansion class, we have an enum to select between how the atom smearings are implemented (constant, radially-dependent, or species-dependent?). 

Another thing to consider here is hidden assumptions: how do you know if the code does not implicitly rely on constant atom smearing, since constant smearing is all that was ever tested?

Reply:
> I think we agree on the goals, but not on how to achieve them. The code already feels very complex, and making it simpler would enable more users to play with it, without them needing to be C++ experts.

Of course I agree that the code is already very complex and daunting for anyone who might want to implement a new representation. Though at the moment, we care more about implementing all the descriptor features we're developing and providing all the functionality people have been expecting in QUIP, rather than making it easy for someone who hasn't seen the code to implement a new representation. There's not many ways to simplify the code that (1) don't compromise the flexibility we've decided from the beginning that we want to implement and (2) don't require a fundamental redesign of important components of the code, delaying the release even further.

> Another thing to consider here is hidden assumptions: how do you know if the code does not implicitly rely on constant atom smearing, since constant smearing is all that was ever tested?

Because we considered this from the beginning when designing the representation, identifying every place in the underlying mathematical expressions where this thing popped up as a "constant" and replacing it with a function of (radius, species, whatever). Sure, the other options will need to be tested when we implement them, just like any other new functionality we implement. If we'd let this become an implicit assumption when designing the representation in the first place, it would be much harder to change than in the current design, where it's left open to future extension.

Reply:
> There's not many ways to simplify the code that (1) don't compromise the flexibility we've decided from the beginning that we want to implement and (2) don't require a fundamental redesign of important components of the code, delaying the release even further.

Yes, I am not arguing to change the current code, getting something that works well out as soon as possible is very important !!

However, I am arguing that in this specific case, the enum and corresponding code clutter does not hold its weight, especially considering the Interpolator already assumes uniform grids (or at least that's my understanding. @agoscinski do you confirm?)